hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/async/FinallyFunction.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdfs.server.federation.router.async;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+/**
+ * The {@code FinallyFunction} interface represents a function that is used to perform
+ * final actions after an asynchronous operation completes, regardless of whether the
+ * operation was successful or resulted in an exception. This interface is part of
+ * the asynchronous utilities provided by the Hadoop Distributed File System (HDFS)
+ * Federation router.
+ *
+ * <p>A {@code FinallyFunction} is typically used for cleanup or finalization tasks,
+ * such as releasing resources, closing connections, or performing other actions that
+ * need to occur after all other operations have completed.</p>
+ *
+ * <p>An implementation of this interface should define what the final action is. It
+ * takes the result of the asynchronous operation as an argument and returns a new
+ * result, which can be the same as the input result or a modified version of it.</p>
+ *
+ * <p>FinallyFunction is used to implement the following semantics:</p>
+ * <pre>
+ * {@code
+ *    try{
+ *      R res = doAsync(input);
+ *    } catch(...) {
+ *      ...
+ *    } finally {
+ *      // Can use FinallyFunction
+ *      R result = thenApply(res);
+ *    }
+ * }
+ * </pre>
+ *
+ * @param <R> the type of the result of the asynchronous operation
+ */
+public interface FinallyFunction<R> {
```

Comment:
@KeeProMise Sir, it's better to add `@FunctionalInterface` annotation here and other place like here.

Reply:
@hfutatzhanghb Thank you for your careful review,  I will add this later.